### DISCUSSION TITLE
pyright working differently for @dataclass decorator vs dataclass as a function. Is this intended behaviour?

### DISCUSSION THREAD

**Discussion by Mxbonn** [microsoft/pyright#9251] [Q&A]
I am no expert in decorators but I thought the following two snippets do the same, yet the second one gives a pyright error:
```python
from dataclasses import dataclass
@dataclass
class DecoratedClass:
    a: int
    b: int

foo = DecoratedClass(a=1, b=2)

class UndecoratedClass:
    a: int
    b: int

UndecoratedClass = dataclass(UndecoratedClass)
bar = UndecoratedClass(a=1, b=2)
```
No parameter named "a"  (reportCallIssue)
No parameter named "b"  (reportCallIssue)

**Comment by erictraut** [comment]
Yes, this is intended. The `@dataclass` decorator requires special-casing in static type checkers because it accepts arguments (like `frozen` or `init`) that affect type checking. You therefore must apply the decorator directly to a class definition if you want it to work with static type checkers.

I'll also note that `@dataclass`, unlike many class decorators, mutates the class that it is applied to rather than returning a new class.
